### ISSUE TITLE
SMAPE overflows for finite inputs with large magnitudes

### ISSUE THREAD

**Issue by Robertboy18** [Lightning-AI/torchmetrics#3508]
SMAPE can return `NaN` or zero for finite inputs because `abs(preds - target)` and `abs(preds) + abs(target)` overflow before the division.

For example, this returns `NaN` and `0.0` on current master:

```python
import torch
from torchmetrics.functional.regression import symmetric_mean_absolute_percentage_error as smape

p = torch.tensor([60000.0], dtype=torch.float16)
print(smape(p, -p))
print(smape(p, p / 2))

# Expected: 2 and approximately 2/3
```

The same problem occurs with `p = torch.tensor([torch.finfo(dtype).max], dtype=dtype)` for float16, bfloat16, float32, and float64, on both CPU and CUDA. The expected ratios are representable; the intermediate sums/differences are the problem.

Tested with Python 3.12.13, PyTorch 2.13.0+cu130, Linux, and TorchMetrics master (`85f4e1686213c8b0b7a1ca6507a67cf654079e88`). CUDA checks used an A100.

I have a patch that halves only the overflowing pairs before computing the ratio, with tests for the functional and streaming APIs, tiny differences, and gradients. It adds eager-execution overhead, so I'll open it as a draft with the timings for discussion. Can attach a short Lean formalization of the rescaling identity if useful :)
